### PR TITLE
[Bifrost] Initial log-server scaffolding

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[Bifrost] Initial log-server scaffolding

This introduces the most basic building blocks for the log-server. Most of the details will be clearer in upcoming PRs, and to allow us to move incrementally, this stubs out many parts that might not appear necessary or used at the moment.

The notable changes:
- Introduction of log-server section in config file.
- Introducing the rocksdb infra and the `LogStore` abstraction to allow the log-server to run with alternative stores (e.g. memory for testing)
- Introduces a LogStoreMarker with startup validation to detect data loss or misconfiguration.
- Automatic transition of StorageState from `Provisioning` to `ReadWrite` after writing a storage marker (auto self-provisioning on new deployment)

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1905).
* #1912
* #1906
* __->__ #1905

### CODE REVIEW

File: crates/log-server/src/metadata.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        serde_json::from_slice(data.as_ref())
+    }
+
+    pub fn node_id(&self) -> PlainNodeId {
+        self.my_node_id
+    }
+
+    pub fn created_at(&self) -> MillisSinceEpoch {
+        self.created_at
+    }
```
Comment (muhamadazmy) [inline crates/log-server/src/metadata.rs:53]: Any reason you prefer getters over making the fields public ?

File: crates/log-server/src/rocksdb_logstore/keys.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    pub fn encode(self, buf: &mut BytesMut) {
+        buf.reserve(Self::serialized_size());
+        buf.put_u8(DATA_KEY_PREFIX);
+        buf.put_u64(*self.loglet_id);
+        // despite offset being 32bit, for backward compatibility and future-proofing, we keep the
+        // storage layer as u64 but we assert that the actual value fits u32 on read.
+        buf.put_u64(self.offset.into());
+    }
```
Comment (muhamadazmy) [inline crates/log-server/src/rocksdb_logstore/keys.rs:72]: This is a not a platform independent encoding. since it has no regard for endianness .

**Review by muhamadazmy — APPROVED** [review]
LGTM. I only have 1 or 2 questions (commented)

File: crates/log-server/src/rocksdb_logstore/keys.rs (hunk 2 above)
Comment (tillrohrmann) [reply crates/log-server/src/rocksdb_logstore/keys.rs:72]: I think it writes the u64 in big endian byte order.

File: crates/log-server/src/rocksdb_logstore/builder.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    opts
+}
+
+fn cf_data_options(
+    memory_budget: usize,
+) -> impl Fn(rocksdb::Options) -> rocksdb::Options + Send + Sync + 'static {
+    move |mut opts| {
+        // memory budget is in bytes. We divide the budget between the data cf and metadata cf.
+        // data 10% to metadata 90% to data.
+        let memtables_budget = (memory_budget as f64 * DATA_CF_BUDGET_RATIO).floor() as usize;
```
Comment (tillrohrmann) [inline crates/log-server/src/rocksdb_logstore/builder.rs:118]: Is `DATA_CF_BUDGET_RATIO` the ratio? If yes, then it is currently set to 85%. Maybe we don't specify the ratio in the comment.

File: crates/log-server/src/rocksdb_logstore/builder.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            DBCompressionType::Lz4,
+            DBCompressionType::Lz4,
+            DBCompressionType::Zstd,
+        ]);
+
+        opts.set_prefix_extractor(SliceTransform::create_fixed_prefix(DATA_KEY_PREFIX_LENGTH));
+        opts.set_memtable_prefix_bloom_ratio(0.2);
+        // most reads are sequential
+        opts.set_advise_random_on_open(false);
+        //
```
Comment (tillrohrmann) [inline crates/log-server/src/rocksdb_logstore/builder.rs:142]: Leftover?

File: crates/log-server/src/rocksdb_logstore/keys.rs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    fn test_metadata_key() {
+        let key = MetadataKey::new(1.into(), MetadataKind::LogletState);
+        assert_eq!(*key.loglet_id, 1);
+        assert_eq!(key.kind, MetadataKind::LogletState);
+        let mut buf = BytesMut::new();
+        let bytes = key.encode_and_split(&mut buf);
+        let key2 = MetadataKey::from_slice(&bytes);
+        assert_eq!(key, key2);
+        assert_eq!(*key2.loglet_id, 1);
+        assert_eq!(key2.kind, MetadataKind::LogletState);
```
Comment (tillrohrmann) [inline crates/log-server/src/rocksdb_logstore/keys.rs:168]: Aren't those checks covered by `assert_eq(key, key2)`?

File: crates/log-server/src/rocksdb_logstore/writer.rs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+            rocksdb,
+            batch_acks_buf: Vec::default(),
+            buffer: BytesMut::with_capacity(INITIAL_SERDE_BUFFER_SIZE),
+            updateable_options,
+        }
+    }
+
+    /// Must be called from task_center context
+    pub fn start(mut self, tc: &TaskCenter) -> Result<RocksDbLogWriterHandle, ShutdownError> {
+        // big enough to allows a second full batch to queue up while the existing one is being processed
```
Comment (tillrohrmann) [inline crates/log-server/src/rocksdb_logstore/writer.rs:58]: s/allows/allow/`

File: crates/types/src/config/log_server.rs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+impl LogServerOptions {
+    pub fn apply_common(&mut self, common: &CommonOptions) {
+        self.rocksdb.apply_common(&common.rocksdb);
+        if self.rocksdb_memory_budget.is_none() {
+            self.rocksdb_memory_budget = Some(
+                // 1MB minimum
+                NonZeroUsize::new(
+                    (common.rocksdb_safe_total_memtables_size() as f64
+                        * self.rocksdb_memory_ratio as f64)
```
Comment (tillrohrmann) [inline crates/types/src/config/log_server.rs:76]: Is `common.rocksdb_safe_total_memtables_size()` shared between `MetadataStore`, `PartitionStore` and `LogServer`? If it is shared, we might go beyond our memory limit.

**Review by tillrohrmann — APPROVED** [review]
Great work @AhmedSoliman :-) LGTM. I only had a small comment regarding the memory configuration (slightly unrelated). +1 for merging.

File: crates/types/src/config/log_server.rs (hunk 7 above)
Comment (AhmedSoliman) [reply crates/types/src/config/log_server.rs:76]: Yes. We might if ratios add to more than 1, but default should be _sane enough_ (0.01+0.49+0.5)

**Pull request merged**